### PR TITLE
libpfm: Fix windows build

### PR DESCRIPTION
**Pull request by t4ccer**
## Description of changes

I've tested `nix-build -A libpfm -A pkgsCross.mingw32.libpfm -A pkgsStatic.libpfm`, can't run `nixpkgs-review` unfortunately, too overwhelming for my laptop

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -23507,7 +23507,9 @@ with pkgs;
 
   libpaper = callPackage ../development/libraries/libpaper { };
 
-  libpfm = callPackage ../development/libraries/libpfm { };
+  libpfm = callPackage ../development/libraries/libpfm {
+    inherit (windows) libgnurx;
```
Comment (Artturin) [inline pkgs/top-level/all-packages.nix:23511]: Use windows directly instead of inheriting

File: pkgs/development/libraries/libpfm/default.nix [hunk 2]
```diff
@@ -1,24 +1,34 @@
...
+  # This target is created only if `ifeq ($(SYS),Linux)` passes
+  patches = [ ./fix-windows.patch ];
+
   makeFlags = [
     "PREFIX=${placeholder "out"}"
     "LDCONFIG=true"
     "ARCH=${stdenv.hostPlatform.uname.processor}"
     "SYS=${stdenv.hostPlatform.uname.system}"
-  ];
+  ] ++ lib.optional stdenv.hostPlatform.isWindows "SYS=WINDOWS"; # It must be capitalized
```
Comment (Artturin) [inline pkgs/development/libraries/libpfm/default.nix:26]: Is the `uname  -s` capitalized on windows, is upstream wrong

File: pkgs/development/libraries/libpfm/default.nix [hunk 3]
```diff
@@ -1,24 +1,34 @@
...
 
   src = fetchurl {
-    url = "mirror://sourceforge/perfmon2/libpfm4/${pname}-${version}.tar.gz";
+    url = "mirror://sourceforge/perfmon2/libpfm4/libpfm-${finalAttrs.version}.tar.gz";
     sha256 = "sha256-0YuXdkx1VSjBBR03bjNUXQ62DG6/hWgENoE/pbBMw9E=";
   };
 
+  # Don't install libpfm.so on windows as it doesn't exist
+  # This target is created only if `ifeq ($(SYS),Linux)` passes
+  patches = [ ./fix-windows.patch ];
```
Comment (Artturin) [inline pkgs/development/libraries/libpfm/default.nix:19]: Upstream?

File: pkgs/development/libraries/libpfm/default.nix (hunk 2 above)
Comment (t4ccer) [reply pkgs/development/libraries/libpfm/default.nix:26]: `pkgsCross.mingw32.stdenv.hostPlatform.uname.system` gives `"Windows"`, but upstream expects `"WINDOWS"`, see https://github.com/wcohen/libpfm4/blob/70b5b4c82912471b43c7ddf0d1e450c4e0ef477e/config.mk#L251-L253 (mirror)

File: pkgs/development/libraries/libpfm/default.nix (hunk 3 above)
Comment (t4ccer) [reply pkgs/development/libraries/libpfm/default.nix:19]: Yeah, maybe worth opening an issue. It's a bit subtle though. [It tries to install SLIBPFM](https://github.com/wcohen/libpfm4/blob/70b5b4c82912471b43c7ddf0d1e450c4e0ef477e/lib/Makefile#L432), but `SLIBPFM` is [set only if we're on linux](https://github.com/wcohen/libpfm4/blob/70b5b4c82912471b43c7ddf0d1e450c4e0ef477e/lib/Makefile#L223-L228)

File: pkgs/development/libraries/libpfm/default.nix (hunk 2 above)
Comment (t4ccer) [reply pkgs/development/libraries/libpfm/default.nix:26]: It seems like we're not hitting [this](https://github.com/wcohen/libpfm4/blob/70b5b4c82912471b43c7ddf0d1e450c4e0ef477e/config.mk#L85-L87) for some reason

Comment (Artturin) [reply pkgs/development/libraries/libpfm/default.nix:26]: https://github.com/git/git/blob/d814540bb75bbd2257f9a6bf59661a84fe8cf3cf/config.mak.uname#L414

has 
Windows
MINGW

so upstream is wrong, it should be matching `Windows`.

Comment (Artturin) [reply pkgs/development/libraries/libpfm/default.nix:26]: Can you substituteInPlace WINDOWS to Windows and add a comment that upstream is wrong.

Comment (t4ccer) [reply pkgs/development/libraries/libpfm/default.nix:26]: Sure

**Pull request merged**